shared/src/main/scala/mlscript/TypeDefs.scala

```diff
@@ -541,5 +541,136 @@ class TypeDefs extends ConstraintSolver { self: Typer =>
     typeMethods(typeTypeDefs(ctx.copy(env = allEnv, mthEnv = allMthEnv, tyDefs = allDefs)))
   }
   
-  
+  /**
+    * Finds variances of all type variables in the given type definitions with the given
+    * context. It starts each type defintion as a bivariant and then keeps changing
+    * based on the type info it encounters. Once it reaches a type variable position
+    * it updates the variance info of the type variable with the variance at that position.
+    * If there is no change then it stops.
+    *
+    * @param tyDefs
+    * @param ctx
+    * @return Mapping from type variable to variance info
+    */
+  def computeVariances(tyDefs: List[TypeDef], ctx: Ctx): VarianceStore = {
+    var varianceUpdated: Bool = false;
+    def updateVariance(ty: SimpleType, curVariance: VarianceInfo)(implicit store: VarianceStore, ctx: Ctx): Unit = ty match {
+      case NegTrait(tt) => ???
+      case NegVar(tv) => ???
```

Comment:
These are just proxies for negated tags and vars. You can remove the cases and match on `ProxyType(underlying)` instead, recursing into `underlying`.